File: phases/10-databases-and-storage/05-physical-storage-pages-slotted-pages/code/main.c
/*
 * Physical Storage -- Pages, Slotted Pages
 * Phase 10 -- Databases & Storage Systems
 *
 * Compile: gcc -Wall -Wextra -o main main.c
 * Run:     ./main
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <stdbool.h>

#define PAGE_SIZE 4096
#define HEADER_SIZE 24
#define SLOT_ENTRY_SIZE 4

/* ---- helpers: unaligned little-endian I/O ---- */

static uint16_t read_u16(const uint8_t *buf, int off) {
    return (uint16_t)buf[off] | ((uint16_t)buf[off + 1] << 8);
}

static void write_u16(uint8_t *buf, int off, uint16_t v) {
    buf[off]     = v & 0xFF;
    buf[off + 1] = (v >> 8) & 0xFF;
}

static uint32_t read_u32(const uint8_t *buf, int off) {
    return (uint32_t)buf[off]       | ((uint32_t)buf[off + 1] << 8) |
           ((uint32_t)buf[off + 2] << 16) | ((uint32_t)buf[off + 3] << 24);
}

static void write_u32(uint8_t *buf, int off, uint32_t v) {
    buf[off]     = v & 0xFF;
    buf[off + 1] = (v >> 8) & 0xFF;
    buf[off + 2] = (v >> 16) & 0xFF;
    buf[off + 3] = (v >> 24) & 0xFF;
}

/* ---- header offsets ---- */
enum { OFF_PAGE_ID = 0, OFF_FREE_START = 4, OFF_DATA_END = 6, OFF_SLOT_COUNT = 8 };

/* ---- SlottedPage ---- */

typedef struct {
    uint8_t buffer[PAGE_SIZE];
} SlottedPage;

void slotted_page_init(SlottedPage *page, uint32_t page_id) {
    memset(page->buffer, 0, PAGE_SIZE);
    write_u32(page->buffer, OFF_PAGE_ID, page_id);
    write_u16(page->buffer, OFF_FREE_START, HEADER_SIZE);
    write_u16(page->buffer, OFF_DATA_END, PAGE_SIZE);
    write_u16(page->buffer, OFF_SLOT_COUNT, 0);
}

uint32_t slotted_page_id(const SlottedPage *page) {
    return read_u32(page->buffer, OFF_PAGE_ID);
}

static uint16_t free_start(const SlottedPage *page) {
    return read_u16(page->buffer, OFF_FREE_START);
}
static void set_free_start(SlottedPage *page, uint16_t v) {
    write_u16(page->buffer, OFF_FREE_START, v);
}

static uint16_t data_end(const SlottedPage *page) {
    return read_u16(page->buffer, OFF_DATA_END);
}
static void set_data_end(SlottedPage *page, uint16_t v) {
    write_u16(page->buffer, OFF_DATA_END, v);
}

static uint16_t slot_count(const SlottedPage *page) {
    return read_u16(page->buffer, OFF_SLOT_COUNT);
}
static void set_slot_count(SlottedPage *page, uint16_t v) {
    write_u16(page->buffer, OFF_SLOT_COUNT, v);
}

static int free_space(const SlottedPage *page) {
    return (int)data_end(page) - (int)free_start(page);
}

/* ---- slot array accessors ---- */

static int slot_entry_off(int slot) {
    return HEADER_SIZE + slot * SLOT_ENTRY_SIZE;
}

static uint16_t slot_off(const SlottedPage *page, int slot) {
    return read_u16(page->buffer, slot_entry_off(slot));
}
static void set_slot_off(SlottedPage *page, int slot, uint16_t v) {
    write_u16(page->buffer, slot_entry_off(slot), v);
}

static uint16_t slot_len(const SlottedPage *page, int slot) {
    return read_u16(page->buffer, slot_entry_off(slot) + 2);
}
static void set_slot_len(SlottedPage *page, int slot, uint16_t v) {
    write_u16(page->buffer, slot_entry_off(slot) + 2, v);
}

static bool slot_is_deleted(const SlottedPage *page, int slot) {
    return slot_off(page, slot) == 0 && slot_len(page, slot) == 0;
}

/* ---- forward declaration ---- */
void slotted_page_defrag(SlottedPage *page);

/* ---- public API ---- */

int slotted_page_insert(SlottedPage *page, const uint8_t *data, int len) {
    int needed = len + SLOT_ENTRY_SIZE;
    if (needed > free_space(page)) {
        slotted_page_defrag(page);
        if (needed > free_space(page)) return -1;
    }
    uint16_t slot = slot_count(page);
    uint16_t new_de = data_end(page) - (uint16_t)len;
    memcpy(page->buffer + new_de, data, (size_t)len);
    set_slot_off(page, slot, new_de);
    set_slot_len(page, slot, (uint16_t)len);
    set_data_end(page, new_de);
    set_slot_count(page, slot + 1);
    set_free_start(page, (uint16_t)(HEADER_SIZE + slot_count(page) * SLOT_ENTRY_SIZE));
    return (int)slot;
}

const uint8_t *slotted_page_get(const SlottedPage *page, int slot, int *out_len) {
    if (slot < 0 || slot >= (int)slot_count(page)) { *out_len = 0; return NULL; }
    if (slot_is_deleted(page, slot)) { *out_len = 0; return NULL; }
    *out_len = (int)slot_len(page, slot);
    return page->buffer + slot_off(page, slot);
}

void slotted_page_delete(SlottedPage *page, int slot) {
    if (slot < 0 || slot >= (int)slot_count(page)) return;
    set_slot_off(page, slot, 0);
    set_slot_len(page, slot, 0);
}

int slotted_page_update(SlottedPage *page, int slot, const uint8_t *data, int len) {
    if (slot < 0 || slot >= (int)slot_count(page)) return -1;
    if (slot_is_deleted(page, slot)) return -1;

    uint16_t old_off = slot_off(page, slot);
    uint16_t old_len = slot_len(page, slot);

    if (len <= (int)old_len) {
        memcpy(page->buffer + old_off, data, (size_t)len);
        if (len < (int)old_len)
            memset(page->buffer + old_off + len, 0, (size_t)(old_len - len));
        set_slot_len(page, slot, (uint16_t)len);
        return 0;
    }

    /* Grow: delete old slot, relocate to end */
    set_slot_off(page, slot, 0);
    set_slot_len(page, slot, 0);

    int needed = len + SLOT_ENTRY_SIZE;
    if (needed > free_space(page)) {
        slotted_page_defrag(page);
        if (needed > free_space(page)) return -1;
    }
    uint16_t new_de = data_end(page) - (uint16_t)len;
    memcpy(page->buffer + new_de, data, (size_t)len);
    set_slot_off(page, slot, new_de);
    set_slot_len(page, slot, (uint16_t)len);
    set_data_end(page, new_de);
    return 0;
}

void slotted_page_defrag(SlottedPage *page) {
    int count = (int)slot_count(page);
    if (count == 0) {
        set_data_end(page, PAGE_SIZE);
        return;
    }
    uint8_t *tmp = (uint8_t *)malloc(PAGE_SIZE);
    int *lens = (int *)malloc((size_t)count * sizeof(int));
    if (!tmp || !lens) { free(tmp); free(lens); return; }

    int toff = 0;
    for (int i = 0; i < count; i++) {
        if (slot_is_deleted(page, i)) {
            lens[i] = 0;
        } else {
            int l = (int)slot_len(page, i);
            memcpy(tmp + toff, page->buffer + slot_off(page, i), (size_t)l);
            lens[i] = l;
            toff += l;
        }
    }

    set_data_end(page, PAGE_SIZE);
    toff = 0;
    for (int i = 0; i < count; i++) {
        if (lens[i] == 0) continue;
        uint16_t no = data_end(page) - (uint16_t)lens[i];
        memcpy(page->buffer + no, tmp + toff, (size_t)lens[i]);
        set_slot_off(page, i, no);
        set_slot_len(page, i, (uint16_t)lens[i]);
        set_data_end(page, no);
        toff += lens[i];
    }
    free(tmp);
    free(lens);
}

/* ---- tests ---- */

static int nfail = 0;
#define TEST(n)  do { printf("  %s ... ", n); } while(0)
#define CHECK(c, m) do { if (!(c)) { printf("FAIL: %s\n", m); nfail++; return; } } while(0)

static void test_insert_and_get(void) {
    SlottedPage p;
    slotted_page_init(&p, 42);
    int s = slotted_page_insert(&p, (const uint8_t *)"hello", 5);
    CHECK(s == 0, "first slot");
    int len;
    const uint8_t *d = slotted_page_get(&p, s, &len);
    CHECK(len == 5, "length");
    CHECK(memcmp(d, "hello", 5) == 0, "content");
    CHECK(slotted_page_id(&p) == 42, "page id");
}

static void test_delete(void) {
    SlottedPage p;
    slotted_page_init(&p, 0);
    int s = slotted_page_insert(&p, (const uint8_t *)"hello", 5);
    slotted_page_delete(&p, s);
    int len;
    CHECK(slotted_page_get(&p, s, &len) == NULL, "deleted gone");
}

static void test_update_shrink(void) {
    SlottedPage p;
    slotted_page_init(&p, 0);
    int s = slotted_page_insert(&p, (const uint8_t *)"longer", 6);
    CHECK(slotted_page_update(&p, s, (const uint8_t *)"hi", 2) == 0, "shrink ok");
    int len; const uint8_t *d = slotted_page_get(&p, s, &len);
    CHECK(len == 2, "shrink len"); CHECK(memcmp(d, "hi", 2) == 0, "shrink content");
}

static void test_update_grow(void) {
    SlottedPage p;
    slotted_page_init(&p, 0);
    int s = slotted_page_insert(&p, (const uint8_t *)"hi", 2);
    CHECK(slotted_page_update(&p, s, (const uint8_t *)"longer data", 11) == 0, "grow ok");
    int len; const uint8_t *d = slotted_page_get(&p, s, &len);
    CHECK(len == 11, "grow len"); CHECK(memcmp(d, "longer data", 11) == 0, "grow content");
}

static void test_defrag(void) {
    SlottedPage p;
    slotted_page_init(&p, 0);
    int s0 = slotted_page_insert(&p, (const uint8_t *)"aaa", 3);
    int s1 = slotted_page_insert(&p, (const uint8_t *)"bbb", 3);
    int s2 = slotted_page_insert(&p, (const uint8_t *)"ccc", 3);
    slotted_page_delete(&p, s1);
    slotted_page_defrag(&p);
    int len;
    const uint8_t *d = slotted_page_get(&p, s0, &len);
    CHECK(len == 3 && memcmp(d, "aaa", 3) == 0, "s0 ok");
    CHECK(slotted_page_get(&p, s1, &len) == NULL, "s1 deleted");
    d = slotted_page_get(&p, s2, &len);
    CHECK(len == 3 && memcmp(d, "ccc", 3) == 0, "s2 ok");
}

static void test_page_full(void) {
    SlottedPage p;
    slotted_page_init(&p, 0);
    uint8_t *big = (uint8_t *)malloc(PAGE_SIZE);
    memset(big, 0, PAGE_SIZE);
    CHECK(slotted_page_insert(&p, big, PAGE_SIZE - HEADER_SIZE - SLOT_ENTRY_SIZE) >= 0, "big fits");
    CHECK(slotted_page_insert(&p, (const uint8_t *)"x", 1) < 0, "overflow fails");
    free(big);
}

static void test_utilization(void) {
    SlottedPage p;
    slotted_page_init(&p, 0);
    double u1 = 1.0 - (double)free_space(&p) / (double)PAGE_SIZE;
    CHECK(u1 > 0.0, "empty page util");
    (void)u1;
}

int main(void) {
    printf("Slotted Page tests:\n");
    test_insert_and_get();
    test_delete();
    test_update_shrink();
    test_update_grow();
    test_defrag();
    test_page_full();
    test_utilization();
    if (nfail == 0) printf("All tests PASSED.\n");
    else printf("%d test(s) FAILED.\n", nfail);
    return nfail > 0 ? 1 : 0;
}

File: phases/10-databases-and-storage/05-physical-storage-pages-slotted-pages/code/main.rs
//! Physical Storage — Pages, Slotted Pages
//! Phase 10 — Databases & Storage Systems
//!
//! Run:   cargo build && cargo run
//! Test:  cargo test

const PAGE_SIZE: usize = 4096;
const HEADER_SIZE: usize = 24;
const SLOT_ENTRY_SIZE: usize = 4;

fn read_u16(buf: &[u8], offset: usize) -> u16 {
    u16::from_le_bytes([buf[offset], buf[offset + 1]])
}

fn write_u16(buf: &mut [u8], offset: usize, val: u16) {
    let b = val.to_le_bytes();
    buf[offset] = b[0];
    buf[offset + 1] = b[1];
}

fn read_u32(buf: &[u8], offset: usize) -> u32 {
    u32::from_le_bytes([buf[offset], buf[offset + 1], buf[offset + 2], buf[offset + 3]])
}

fn write_u32(buf: &mut [u8], offset: usize, val: u32) {
    let b = val.to_le_bytes();
    buf[offset] = b[0];
    buf[offset + 1] = b[1];
    buf[offset + 2] = b[2];
    buf[offset + 3] = b[3];
}

/// A slotted page — fixed-size buffer with variable-length records.
///
/// Layout:
/// ```text
/// [Header 24B] [Slot Array → ... ] ← Free Space → ... [← Record Data] [← PAGE_SIZE]
/// ```
///
/// Header fields:
///   [0..4)  page_id      (u32)
///   [4..6)  free_start   (u16)  — first byte of free space (end of slot array)
///   [6..8)  data_end     (u16)  — first byte before record data (grows down)
///   [8..10) slot_count   (u16)
///   [10..24) reserved
///
/// Slot entry (4 bytes per slot):
///   offset (u16) — byte offset of record data in the buffer
///   length (u16) — byte length of record data
///   An entry of (0, 0) indicates a deleted slot.
pub struct SlottedPage {
    buffer: [u8; PAGE_SIZE],
}

impl SlottedPage {
    pub fn new(page_id: u32) -> Self {
        let mut buffer = [0u8; PAGE_SIZE];
        write_u32(&mut buffer, 0, page_id);
        write_u16(&mut buffer, 4, HEADER_SIZE as u16);
        write_u16(&mut buffer, 6, PAGE_SIZE as u16);
        SlottedPage { buffer }
    }

    // --- header accessors ---

    fn free_start(&self) -> u16 {
        read_u16(&self.buffer, 4)
    }
    fn set_free_start(&mut self, v: u16) {
        write_u16(&mut self.buffer, 4, v);
    }
    fn data_end(&self) -> u16 {
        read_u16(&self.buffer, 6)
    }
    fn set_data_end(&mut self, v: u16) {
        write_u16(&mut self.buffer, 6, v);
    }
    fn slot_count(&self) -> u16 {
        read_u16(&self.buffer, 8)
    }
    fn set_slot_count(&mut self, v: u16) {
        write_u16(&mut self.buffer, 8, v);
    }

    fn free_space(&self) -> usize {
        self.data_end() as usize - self.free_start() as usize
    }

    fn slot_entry_base(&self, slot: u16) -> usize {
        HEADER_SIZE + slot as usize * SLOT_ENTRY_SIZE
    }
    fn slot_off(&self, slot: u16) -> u16 {
        let b = self.slot_entry_base(slot);
        read_u16(&self.buffer, b)
    }
    fn set_slot_off(&mut self, slot: u16, v: u16) {
        let b = self.slot_entry_base(slot);
        write_u16(&mut self.buffer, b, v);
    }
    fn slot_len(&self, slot: u16) -> u16 {
        let b = self.slot_entry_base(slot);
        read_u16(&self.buffer, b + 2)
    }
    fn set_slot_len(&mut self, slot: u16, v: u16) {
        let b = self.slot_entry_base(slot);
        write_u16(&mut self.buffer, b + 2, v);
    }

    // --- public API ---

    pub fn page_id(&self) -> u32 {
        read_u32(&self.buffer, 0)
    }

    pub fn utilization(&self) -> f64 {
        let used = PAGE_SIZE - self.free_space();
        used as f64 / PAGE_SIZE as f64
    }

    pub fn insert_record(&mut self, data: &[u8]) -> Result<u16, &str> {
        let needed = data.len() + SLOT_ENTRY_SIZE;
        if needed > self.free_space() {
            self.defragment();
            if needed > self.free_space() {
                return Err("page full");
            }
        }
        let slot = self.slot_count();
        let new_de = self.data_end() - data.len() as u16;
        self.buffer[new_de as usize..][..data.len()].copy_from_slice(data);
        self.set_slot_off(slot, new_de);
        self.set_slot_len(slot, data.len() as u16);
        self.set_data_end(new_de);
        self.set_slot_count(slot + 1);
        self.set_free_start(HEADER_SIZE as u16 + self.slot_count() * SLOT_ENTRY_SIZE as u16);
        Ok(slot)
    }

    pub fn get_record(&self, slot: u16) -> Option<&[u8]> {
        if slot >= self.slot_count() {
            return None;
        }
        let off = self.slot_off(slot);
        let len = self.slot_len(slot);
        if off == 0 && len == 0 {
            return None;
        }
        Some(&self.buffer[off as usize..][..len as usize])
    }

    pub fn delete_record(&mut self, slot: u16) {
        if slot >= self.slot_count() {
            return;
        }
        self.set_slot_off(slot, 0);
        self.set_slot_len(slot, 0);
    }

    pub fn update_record(&mut self, slot: u16, data: &[u8]) -> Result<(), &str> {
        if slot >= self.slot_count() {
            return Err("invalid slot");
        }
        let old_off = self.slot_off(slot);
        let old_len = self.slot_len(slot);
        if old_off == 0 && old_len == 0 {
            return Err("slot is deleted");
        }
        // Shrink: copy in place, update length
        if data.len() <= old_len as usize {
            let start = old_off as usize;
            self.buffer[start..][..data.len()].copy_from_slice(data);
            if data.len() < old_len as usize {
                self.buffer[start + data.len()..][..old_len as usize - data.len()].fill(0);
            }
            self.set_slot_len(slot, data.len() as u16);
            return Ok(());
        }
        // Grow: relocate to end
        self.set_slot_off(slot, 0);
        self.set_slot_len(slot, 0);
        let needed = data.len() + SLOT_ENTRY_SIZE;
        if needed > self.free_space() {
            self.defragment();
            if needed > self.free_space() {
                return Err("page full after defrag");
            }
        }
        let new_de = self.data_end() - data.len() as u16;
        self.buffer[new_de as usize..][..data.len()].copy_from_slice(data);
        self.set_slot_off(slot, new_de);
        self.set_slot_len(slot, data.len() as u16);
        self.set_data_end(new_de);
        Ok(())
    }

    pub fn defragment(&mut self) {
        let count = self.slot_count() as usize;
        if count == 0 {
            self.set_data_end(PAGE_SIZE as u16);
            return;
        }
        // Collect live records
        struct Rec {
            slot: usize,
            data: Vec<u8>,
        }
        let mut live: Vec<Rec> = Vec::with_capacity(count);
        for i in 0..count {
            let off = self.slot_off(i as u16);
            let len = self.slot_len(i as u16);
            if off != 0 || len != 0 {
                let slice = &self.buffer[off as usize..][..len as usize];
                live.push(Rec { slot: i, data: slice.to_vec() });
            }
        }
        self.set_data_end(PAGE_SIZE as u16);
        for rec in &live {
            let new_off = self.data_end() - rec.data.len() as u16;
            self.buffer[new_off as usize..][..rec.data.len()].copy_from_slice(&rec.data);
            self.set_slot_off(rec.slot as u16, new_off);
            self.set_slot_len(rec.slot as u16, rec.data.len() as u16);
            self.set_data_end(new_off);
        }
    }
}

/// A heap file — flat collection of slotted pages.
/// Allocates new pages when existing pages are full.
pub struct HeapFile {
    pages: Vec<SlottedPage>,
    next_page_id: u32,
}

impl HeapFile {
    pub fn new() -> Self {
        HeapFile { pages: Vec::new(), next_page_id: 0 }
    }

    pub fn insert_record(&mut self, data: &[u8]) -> (u32, u16) {
        for page in self.pages.iter_mut() {
            if let Ok(slot) = page.insert_record(data) {
                return (page.page_id(), slot);
            }
        }
        let mut page = SlottedPage::new(self.next_page_id);
        self.next_page_id += 1;
        let slot = page.insert_record(data).unwrap();
        let pid = page.page_id();
        self.pages.push(page);
        (pid, slot)
    }

    pub fn get_record(&self, pid: u32, slot: u16) -> Option<&[u8]> {
        self.pages.iter().find(|p| p.page_id() == pid)
            .and_then(|p| p.get_record(slot))
    }

    pub fn delete_record(&mut self, pid: u32, slot: u16) {
        if let Some(page) = self.pages.iter_mut().find(|p| p.page_id() == pid) {
            page.delete_record(slot);
        }
    }

    pub fn num_pages(&self) -> usize {
        self.pages.len()
    }
}

fn main() {
    let mut page = SlottedPage::new(1);

    let s0 = page.insert_record(b"hello").unwrap();
    let s1 = page.insert_record(b"world").unwrap();
    let s2 = page.insert_record(b"slotted page").unwrap();

    println!("page_id={} slots={} util={:.2}%",
        page.page_id(), page.slot_count(), page.utilization() * 100.0);

    for s in [s0, s1, s2] {
        if let Some(data) = page.get_record(s) {
            println!("  slot {}: {:?}", s, std::str::from_utf8(data).unwrap());
        }
    }

    page.delete_record(s1);
    println!("after delete slot {}: {:?}", s1, page.get_record(s1));

    page.update_record(s0, b"HELLO AGAIN").unwrap();
    println!("after update slot {}: {:?}", s0, std::str::from_utf8(page.get_record(s0).unwrap()).unwrap());

    page.defragment();
    println!("after defrag: util={:.2}%", page.utilization() * 100.0);
    for s in [s0, s1, s2] {
        if let Some(data) = page.get_record(s) {
            println!("  slot {}: {:?}", s, std::str::from_utf8(data).unwrap());
        }
    }

    let mut hf = HeapFile::new();
    let (pid, _) = hf.insert_record(b"heap file record");
    println!("\nheap file: page_id={}, num_pages={}", pid, hf.num_pages());
    assert_eq!(hf.get_record(pid, 0), Some(&b"heap file record"[..]));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_insert_and_get() {
        let mut p = SlottedPage::new(0);
        let s = p.insert_record(b"hello").unwrap();
        assert_eq!(p.get_record(s), Some(&b"hello"[..]));
    }

    #[test]
    fn test_delete() {
        let mut p = SlottedPage::new(0);
        let s = p.insert_record(b"hello").unwrap();
        p.delete_record(s);
        assert_eq!(p.get_record(s), None);
    }

    #[test]
    fn test_update_shrink() {
        let mut p = SlottedPage::new(0);
        let s = p.insert_record(b"longer data").unwrap();
        p.update_record(s, b"hi").unwrap();
        assert_eq!(p.get_record(s), Some(&b"hi"[..]));
    }

    #[test]
    fn test_update_grow() {
        let mut p = SlottedPage::new(0);
        let s = p.insert_record(b"hi").unwrap();
        p.update_record(s, b"longer data here").unwrap();
        assert_eq!(p.get_record(s), Some(&b"longer data here"[..]));
    }

    #[test]
    fn test_defrag_leaves_slot_numbers_stable() {
        let mut p = SlottedPage::new(0);
        let s0 = p.insert_record(b"aaa").unwrap();
        let s1 = p.insert_record(b"bbb").unwrap();
        let s2 = p.insert_record(b"ccc").unwrap();
        p.delete_record(s1);
        p.defragment();
        assert_eq!(p.get_record(s0), Some(&b"aaa"[..]));
        assert_eq!(p.get_record(s1), None);
        assert_eq!(p.get_record(s2), Some(&b"ccc"[..]));
    }

    #[test]
    fn test_page_full() {
        let mut p = SlottedPage::new(0);
        let big = vec![0u8; PAGE_SIZE - HEADER_SIZE - SLOT_ENTRY_SIZE];
        assert!(p.insert_record(&big).is_ok());
        assert!(p.insert_record(b"x").is_err());
    }

    #[test]
    fn test_heap_file_spills() {
        let mut hf = HeapFile::new();
        let big = vec![0u8; PAGE_SIZE - HEADER_SIZE - SLOT_ENTRY_SIZE];
        let (p1, _) = hf.insert_record(&big);
        let (p2, _) = hf.insert_record(b"spill");
        assert_ne!(p1, p2);
        assert_eq!(hf.num_pages(), 2);
    }

    #[test]
    fn test_utilization_nonzero() {
        let mut p = SlottedPage::new(0);
        assert!(p.utilization() > 0.0);
        let _ = p.insert_record(b"data");
        let u = p.utilization();
        assert!(u > 0.0 && u < 1.0);
    }
}
